tests/src/test/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManagerSuite.scala

```diff
@@ -0,0 +1,130 @@
+    context
+  }
+
+  test("test basic non-blocking requests") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestNonBlockingLease(context, 100)) { lease =>
+      assert(lease.leaseAmount == 100)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 100)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+  }
+
+  test("test shrinking non-blocking request") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestNonBlockingLease(context, 2048)) { lease =>
+      assert(lease.leaseAmount == 1024)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 1024)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test blocking call that does not block") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    withResource(GpuMemoryLeaseManager.requestLease(context, 100)) { lease =>
+      assert(lease.leaseAmount == 100)
+      assert(GpuMemoryLeaseManager.getTotalLease(context) == 100)
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test blocking too large required throws") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val context = mockContext(1)
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    assertThrows[IllegalArgumentException] {
+      withResource(GpuMemoryLeaseManager.requestLease(context, 0, 2048)) { _ =>
+      }
+    }
+    assert(GpuMemoryLeaseManager.getTotalLease(context) == 0)
+    GpuMemoryLeaseManager.releaseAllForTask(context)
+  }
+
+  test("test basic blocking behavior") {
+    GpuMemoryLeaseManager.initializeForTest(1024)
+    val firstContext = mockContext(1)
+    val secondContext = mockContext(2)
+    assert(GpuMemoryLeaseManager.getTotalLease(firstContext) == 0)
+    assert(GpuMemoryLeaseManager.getTotalLease(secondContext) == 0)
+    var childLease: MemoryLease = null
+    var childHasStarted: Boolean = false
+    val child = new Thread() {
+      override def run(): Unit = {
+        childHasStarted = true
+        childLease = GpuMemoryLeaseManager.requestLease(secondContext, 900)
+      }
+    }
+    withResource(GpuMemoryLeaseManager.requestLease(firstContext, 900)) { lease =>
+      assert(lease.leaseAmount == 900)
+      child.start()
+      // The child should block so give it some time to come up...
+      val timeoutTime = System.nanoTime() + TimeUnit.SECONDS.toNanos(10)
```

Comment:
should we make this millis? It seems like this is going to loop exactly once after sleeping for 100ms, if not done within 10 nanos.

Reply:
The java APIs for TimeUnit are really confusing, in my opinion, that is saying to wait for up to 10 seconds converted into nanoseconds.

Reply:
oh sorry I am confused. Thank you